### PR TITLE
Docker generator

### PR DESCRIPTION
**Pull request by sidot3291**
This adds `local-compose gen docker` for cross-compatibility with Docker.

@burrows-labs please review architecture and let me know if I should refactor.  I'm thinking we'll use a similar structure (parseGenSubcommand + genX.go) for future generators.

Thanks

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@ local-compose
 local-compose.exe
 /testblobs/
 /testdir123/
+.DS_Store
```
Comment (burrows-labs) [inline .gitignore]: Add a newline at end of file.

File: genDocker.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	}
+}
+
+func updatePGConf() {
+	// Idempotent.
+	// Clone the behavior from these links so postgres is open to other hosts:
+	// https://github.com/docker-library/postgres/blob/87b15b6c65ba985ac958e7b35ba787422113066e/11/Dockerfile#L160
+	// https://github.com/docker-library/postgres/blob/818958f40191c614b1e17dc9e5249ddbb7406a60/11/alpine/docker-entrypoint.sh#L119
+	pgconf := path.Join(".local-compose", "postgres", "data", "postgresql.conf")
+	input, err := ioutil.ReadFile(pgconf)
```
Comment (burrows-labs) [inline genDocker.go]: Use `readFile`.  Don't let a big file nuke the process.

File: genDocker.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+				lines[i] = "listen_addresses = '*'"
+			}
+		}
+		if writeStringFile(pgconf, strings.Join(lines, "\n")) == false {
+			panic("Failed to write postgresql.conf")
+		}
+	}
+
+	pghba := path.Join(".local-compose", "postgres", "data", "pg_hba.conf")
+	input, err = ioutil.ReadFile(pghba)
```
Comment (burrows-labs) [inline genDocker.go]: `readFile`

File: genDocker.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+		}
+		if !trusted {
+			lines = append(lines, trustLine)
+		}
+
+		if writeStringFile(pghba, strings.Join(lines, "\n")) == false {
+			panic("Failed to write pg_hba.conf")
+		}
+	}
+}
```
Comment (burrows-labs) [inline genDocker.go]: What happens when both files get updated?
What happens when neither file is updated?

File: genDocker.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		for i, line := range lines {
+			if strings.Contains(line, "#listen_addresses = 'localhost'") {
+				lines[i] = "listen_addresses = '*'"
+			}
+		}
+		if writeStringFile(pgconf, strings.Join(lines, "\n")) == false {
+			panic("Failed to write postgresql.conf")
+		}
+	}
+
```
Comment (burrows-labs) [inline genDocker.go]: I would maybe fail here if ReadFile fails for a reason other than "file not found".

File: genNginx.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	nginx = append(nginx, "http {")
+	nginx = append(nginx, "  include            mime.types;")
+	nginx = append(nginx, "  default_type       application/octet-stream;")
+	nginx = append(nginx, "  sendfile           on;")
+	nginx = append(nginx, "  keepalive_timeout  65;")
+	nginx = append(nginx, "  server {")
+	nginx = append(nginx, "    listen 3000;")
+	for dir, x := range projectConfig.Services {
+		nginx = append(nginx, fmt.Sprintf("    location %s {", x.UrlPathPrefix))
+		nginx = append(nginx, fmt.Sprintf("      resolver    127.0.0.11 valid=30s;"))
```
Comment (burrows-labs) [inline genNginx.go]: This is the docker embedded DNS server?

File: genDocker.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+		panic(fmt.Sprintf("Failed to write Dockerfile for %s", dir))
+	}
+}
+
+func updatePGConf() {
+	// Idempotent.
+	// Clone the behavior from these links so postgres is open to other hosts:
+	// https://github.com/docker-library/postgres/blob/87b15b6c65ba985ac958e7b35ba787422113066e/11/Dockerfile#L160
+	// https://github.com/docker-library/postgres/blob/818958f40191c614b1e17dc9e5249ddbb7406a60/11/alpine/docker-entrypoint.sh#L119
+	pgconf := path.Join(".local-compose", "postgres", "data", "postgresql.conf")
```
Comment (burrows-labs) [inline genDocker.go]: Why modify a file outside of docker?

File: genNginx.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+		nginx = append(nginx, fmt.Sprintf("      proxy_set_header  X-Forwarded-Port $server_port;"))
+		nginx = append(nginx, fmt.Sprintf("      proxy_set_header  X-Forwarded-Host $host;"))
+		nginx = append(nginx, fmt.Sprintf("      proxy_set_header  Upgrade $http_upgrade;"))
+		nginx = append(nginx, fmt.Sprintf("      proxy_set_header  Connection \"upgrade\";"))
+		nginx = append(nginx, fmt.Sprintf("      proxy_redirect off;"))
+		nginx = append(nginx, fmt.Sprintf("      proxy_http_version 1.1;"))
+		nginx = append(nginx, fmt.Sprintf("    }"))
+	}
+	nginx = append(nginx, "  }")
+	nginx = append(nginx, "  include servers/*;")
```
Comment (burrows-labs) [inline genNginx.go]: What is this `include servers/*` doing?

File: genDocker.go (hunk 7 above)
Comment (burrows-labs) [inline genDocker.go]: Should docker and not-docker be sharing postgres configs?

File: genNginx.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	nginx = append(nginx, "}")
+	if writeStringFile("nginx.conf", strings.Join(nginx, "\n")) == false {
```
Comment (burrows-labs) [inline genNginx.go]: This nginx file is specific to docker. Maybe call it `nginx-docker.conf` or something.

File: genDocker.go [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+	var dockerCompose []string
+
+	dockerCompose = append(dockerCompose, "version: '3.4'")
+	dockerCompose = append(dockerCompose, "x-database-variables:")
+	dockerCompose = append(dockerCompose, "  &database-variables")
+	dockerCompose = append(dockerCompose, "  DATABASE_URL: postgres://postgres:@postgres/local-compose")
+
+	dockerCompose = append(dockerCompose, "services:")
+	dockerCompose = append(dockerCompose, "  postgres:")
```
Comment (burrows-labs) [inline genDocker.go]: You are always adding the postgres service?

File: genDocker.go [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+	dockerfile = append(dockerfile, "RUN gem install bundler && bundle install")
+	dockerfile = append(dockerfile, "COPY . /app")
+	dockerfile = append(dockerfile, "EXPOSE $PORT")
+
+	if writeStringFile(path.Join(dir, "Dockerfile"), strings.Join(dockerfile, "\n")) == false {
+		panic(fmt.Sprintf("Failed to write Dockerfile for %s", dir))
+	}
+}
+
+func updatePGConf() {
```
Comment (burrows-labs) [inline genDocker.go]: Need better func name.

File: genDocker.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+	"strings"
+)
+
+func genDocker() {
+	projectConfig, b := localComposeProjectGate()
+	if b == false {
+		return
+	}
+
+	updatePGConf()
```
Comment (burrows-labs) [inline genDocker.go]: If the user has never done `local-compose start` then the postgres files won't exist.

File: genDocker.go (hunk 7 above)
Comment (sidot3291) [reply genDocker.go]: In this case I think it's, should local docker development and local-compose be sharing postgres configs?  I'm not expecting this config to make it to production.

Do you concerns with the two changes in updatePGConf being applied to local-compose, too?  Any additional concerns?

In general I like that local-compose and Docker point at the same data folder, but that doesn't stop us from changing the config across environments, if necessary.

File: genNginx.go (hunk 8 above)
Comment (sidot3291) [reply genNginx.go]: Removing

File: genDocker.go [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+
+	dockerCompose = append(dockerCompose, "services:")
+	dockerCompose = append(dockerCompose, "  postgres:")
+	dockerCompose = append(dockerCompose, "    image: postgres:10.9")
+	dockerCompose = append(dockerCompose, "    volumes:")
+	dockerCompose = append(dockerCompose, "      - ./.local-compose/postgres/data:/var/lib/postgresql/data")
+	dockerCompose = append(dockerCompose, "    ports:")
+	dockerCompose = append(dockerCompose, "      - \"20000:5432\"")
+	dockerCompose = append(dockerCompose, "    environment:")
+	dockerCompose = append(dockerCompose, "      - POSTGRES_DB=local-compose")
```
Comment (sidot3291) [inline genDocker.go]: Accidentally deleted your comment, but:

Will change to only start if there's a backend service, mimicking `startPostgresOnceIfNeeded`

File: genNginx.go (hunk 9 above)
Comment (sidot3291) [reply genNginx.go]: Updating.

File: genDocker.go (hunk 11 above)
Comment (sidot3291) [reply genDocker.go]: Changing to `updatePostgresIngressConfig`

File: genDocker.go (hunk 7 above)
Comment (burrows-labs) [reply genDocker.go]: >> In this case I think it's, should local docker development and local-compose be sharing postgres configs?

Yes.

>> Do you concerns with the two changes in updatePGConf being applied to local-compose, too? Any additional concerns?

It's surprising that generating dockerfiles will change the behavior of local-compose's database.

File: genDocker.go (hunk 13 above)
Comment (burrows-labs) [reply genDocker.go]: Do you mean start or adding to config?

Comment (sidot3291) [reply genDocker.go]: Adding to config.

**Comment by sidot3291** [comment]
@burrows-labs this should be ready.  dockers postgres no longer points at l-c's, so a handful of comments disappeared

**Pull request closed without merging**